### PR TITLE
feat(es/typescript): inline enum

### PR DESCRIPTION
**Pull request by magic-akari**
**Description:**
Inline enum if possible.

**BREAKING CHANGE:**

**Related issue (if exists):**

### CODE REVIEW

**Comment by kdy1** [comment]
Does this fix #940?

**Comment by magic-akari** [comment]
Partially.
It does not work across files.

File: crates/swc_ecma_transforms_typescript/src/inline_enum.rs [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+#[serde(rename_all = "camelCase")]
+#[serde(untagged)]
+pub enum InlineEnumConfig {
+    Const(EnumConst),
+    All(bool),
+}
+
+impl Default for InlineEnumConfig {
+    fn default() -> Self {
+        Self::All(true)
```
Comment (magic-akari) [inline crates/swc_ecma_transforms_typescript/src/inline_enum.rs]: @kdy1 
I set "inline all as possible" as default option.
What do you think about it?

Comment (kdy1) [reply crates/swc_ecma_transforms_typescript/src/inline_enum.rs]: I think it's good enough to be default

File: crates/swc_ecma_transforms_typescript/src/inline_enum.rs [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+                    }
+                }
+            }
+        }
+
+        e.visit_mut_children_with(self);
+    }
+
+    fn visit_mut_assign_expr(&mut self, a: &mut AssignExpr) {
+        a.right.visit_mut_with(self);
```
Comment (kdy1) [inline crates/swc_ecma_transforms_typescript/src/inline_enum.rs]: `a.left` can store expressions.

File: crates/swc_ecma_transforms_typescript/src/inline_enum.rs [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+use serde::{Deserialize, Serialize};
+use swc_atoms::JsWord;
+use swc_common::collections::AHashMap;
+use swc_ecma_ast::{
+    AssignExpr, ComputedPropName, Expr, Id, Lit, MemberExpr, MemberProp, Module, Script, Str, Tpl,
+};
+use swc_ecma_utils::ident::IdentLike;
+use swc_ecma_visit::{as_folder, noop_visit_mut_type, Fold, VisitMut, VisitMutWith};
+
+pub type TSEnumLit = Rc<RefCell<AHashMap<Id, AHashMap<JsWord, Lit>>>>;
```
Comment (kdy1) [inline crates/swc_ecma_transforms_typescript/src/inline_enum.rs]: Can you change this to a struct which hides the actual type?
The field can be like

```rust
pub(crate) values: Rc<RefCell<AHashMap<Id, AHashMap<JsWord, Lit>>>>
```

File: crates/swc_ecma_transforms_typescript/src/inline_enum.rs (hunk 1 above)
Comment (magic-akari) [reply crates/swc_ecma_transforms_typescript/src/inline_enum.rs]: There is a risk to make it default.
See https://www.typescriptlang.org/play?#code/KYOwrgtgBAglDeAoKsA0UBCiC+jEAo4BDAZyiJAE8BKAbQHIZ6BdKAXigEYBuA4sijVoAGVh06IgA

Comment (kdy1) [inline crates/swc_ecma_transforms_typescript/src/inline_enum.rs]: If so, we have two options I guess.

The first one is adding a field to `Assumptions` and the second one is defaulting to `EnumConst::Const`.
I think the second one would be better at the moment

**Review by kdy1 — CHANGES_REQUESTED** [review]
Thank you!

File: crates/swc_ecma_transforms_typescript/src/inline_enum.rs (hunk 1 above)
Comment (Austaras) [reply crates/swc_ecma_transforms_typescript/src/inline_enum.rs]: I prefer an assumption, we need to expose `Assumptions` in a near future anyway.

Comment (kdy1) [reply crates/swc_ecma_transforms_typescript/src/inline_enum.rs]: Oh, I see. I think the assumption is good too.

Comment (Austaras) [reply crates/swc_ecma_transforms_typescript/src/inline_enum.rs]: That would also need to change the options from `InlineEnumConfig` to `bool`

File: crates/swc_ecma_transforms_typescript/src/strip.rs [hunk 4]
```diff
@@ -985,15 +1012,33 @@ where
...
+            let mut enums = self.ts_enum_lit.borrow_mut();
+
+            let k_v = enums.entry(e.id.to_id()).or_default();
+
+            members
+                .iter()
+                .filter(|(_, v)| matches!(v, Expr::Lit(..)))
+                .for_each(|(m, e)| {
+                    let member_key = m.id.as_ref().clone();
+                    let value = e.clone().expect_lit();
```
Comment (kdy1) [inline crates/swc_ecma_transforms_typescript/src/strip.rs:1028]: Can you add a test for enum which has a template literal in its initializer?

File: crates/swc_ecma_transforms_typescript/src/inline_enum.rs [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+}
+
+struct InlineEnum {
+    ts_enum_lit: TSEnumLit,
+    ts_enum_config: TSEnumConfig,
+
+    is_lhs: bool,
+}
+
+trait SkipInlineEnum {
```
Comment (kdy1) [inline crates/swc_ecma_transforms_typescript/src/inline_enum.rs:31]: Is this trait required?

File: crates/swc_ecma_transforms_typescript/src/inline_enum.rs [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+    fn skip(&self) -> bool;
+}
+
+impl SkipInlineEnum for InlineEnum {
```
Comment (kdy1) [inline crates/swc_ecma_transforms_typescript/src/inline_enum.rs:35]: Seems like this is the only impl.
I think it can be a method of `InlineEnum`

**Review by kdy1 — CHANGES_REQUESTED** [review]
Thank you!

File: crates/swc_ecma_transforms_typescript/src/inline_enum.rs (hunk 6 above)
Comment (magic-akari) [reply crates/swc_ecma_transforms_typescript/src/inline_enum.rs]: I had implemented this method on another pub struct while I want to hide it, so I wrote this trait.
I am going to fix it.

File: crates/swc_ecma_transforms_typescript/src/strip.rs (hunk 4 above)
Comment (magic-akari) [reply crates/swc_ecma_transforms_typescript/src/strip.rs:1028]: Done.
see:
https://github.com/swc-project/swc/blob/ac1f5f97f0d801608994b84ff7b51a194d3a2174/crates/swc/tests/fixture/ts-inline-enum/default/input/index.ts#L11-L18
https://github.com/swc-project/swc/blob/ac1f5f97f0d801608994b84ff7b51a194d3a2174/crates/swc/tests/fixture/ts-inline-enum/default/output/index.ts#L10-L17

File: crates/swc_ecma_transforms_typescript/src/inline_enum.rs [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+    Module, Script, Str, Tpl,
+};
+use swc_ecma_utils::ident::IdentLike;
+use swc_ecma_visit::{as_folder, noop_visit_mut_type, Fold, VisitMut, VisitMutWith};
+
+use crate::TSEnumConfig;
+
+pub(crate) type TSEnumLit = Rc<RefCell<AHashMap<Id, AHashMap<JsWord, Lit>>>>;
+
+pub fn inline_enum(ts_enum_lit: TSEnumLit, ts_enum_config: TSEnumConfig) -> impl Fold + VisitMut {
```
Comment (kdy1) [inline crates/swc_ecma_transforms_typescript/src/inline_enum.rs:16]: This function is exposing a crate-local type (`TSEnumLit`) in public api.

Comment (kdy1) [reply crates/swc_ecma_transforms_typescript/src/inline_enum.rs:16]: Maybe it should be `pub(crate) fn`?

**Review by kdy1 — CHANGES_REQUESTED** [review]
I found some tiny issues, thank you!

File: crates/swc_ecma_transforms_typescript/src/inline_enum.rs (hunk 7 above)
Comment (magic-akari) [reply crates/swc_ecma_transforms_typescript/src/inline_enum.rs:16]: Yes, `inline_enum` should not be public API

**Review by kdy1 — APPROVED** [review]
Thank you!
LGTM.
Sorry for the delay, I had lots of PRs to review.


---


swc-bump:
 - swc_ecma_transforms_typescript --breaking

**Pull request merged**